python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -0,0 +1,101 @@
+import pathlib
+from typing import TYPE_CHECKING, List, Optional, Union
+
+import numpy as np
+from ray.data.datasource.binary_datasource import BinaryDatasource
+from ray.data.datasource.datasource import Reader
+from ray.data.datasource.file_based_datasource import (
+    _resolve_paths_and_filesystem,
+    FileExtensionFilter,
+)
+from ray.data.datasource.partitioning import PathPartitionFilter
+
+if TYPE_CHECKING:
+    import pyarrow
+    from ray.data.block import T
+
+IMAGE_EXTENSIONS = ["png", "jpg", "jpeg", "tiff", "bmp", "gif"]
+
+
+class ImageFolderDatasource(BinaryDatasource):
+    """A datasource that lets you read datasets like `ImageNet <https://www.image-net.org/>`_."""  # noqa: E501
+
+    def create_reader(
+        self,
+        paths: Union[str, List[str]],
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        partition_filter: PathPartitionFilter = None,
+        **kwargs,
+    ) -> "Reader[T]":
+        if len(paths) != 1:
+            raise ValueError(
+                "`ImageFolderDatasource` expects 1 path representing the dataset "
+                f"root, but it got {len(paths)} paths instead. To fix this "
+                "error, pass in a single-element list containing the dataset root "
+                '(for example, `paths=["s3://imagenet/train"]`)'
+            )
+
+        try:
+            import imageio  # noqa: F401
+        except ImportError:
+            raise ValueError(
+                "`ImageFolderDatasource` depends on 'imageio', but 'imageio' couldn't "
+                "be imported. You can install 'imageio' by running "
+                "`pip install imageio`."
+            )
+
+        if partition_filter is None:
+            partition_filter = FileExtensionFilter(file_extensions=IMAGE_EXTENSIONS)
+
+        # We call `_resolve_paths_and_filesystem` so that the dataset root is formatted
+        # in the same way as the paths passed to `_get_class_from_path`.
+        paths, filesystem = _resolve_paths_and_filesystem(paths, filesystem)
+        self.root = paths[0]
+
+        from pyarrow.fs import FileType, FileSelector
+
+        labels = [
+            file_info.base_name
+            for file_info in filesystem.get_file_info(FileSelector(self.root))
+            if file_info.type is FileType.Directory
+        ]
+        labels.sort()  # Sort labels so that targets are consistent.
+        self.label_to_target = {label: labels.index(label) for label in labels}
```

Comment:
> Is this mapping available as metadata in the Dataset anywhere?

I'm not sure if I totally understand what you mean. Can you give me more information?

Reply:
Ah not an issue anymore since we dropped `target` column. But the original intent was to ask how the user would access `label_to_target` from the created Dataset.